Fix timequery returning wrong offset after trim-prefix which could lead to stuck consumers

nvartolomei opened the pull request:
The fix works only if compression is not in use. We need follow-up work which would decompress the batches to find the exact offset to return, or (!) we could to prevent trim-offset inside a batch in that case.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.1.x
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes

### Bug Fixes

* Fix a scenario where list_offset with a timestamp could return a lower offset than partition start after a trim-prefix command. This could lead to consumers being stuck with an out-of-range-offset exception if they began consuming from an offset below the one which was used in the trim-prefix command.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei commented:
/dt

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/48359#018f1bfc-2f63-4411-a634-cf0b04ce2121:
```
"rptest.tests.timequery_test.TimeQueryTest.test_timequery_with_trim_prefix.cloud_storage=True.spillover=True"
```
<!--857481b6-aa55-4f8f-a2b3-8b32b893a417-->

nvartolomei commented:
/dt

andijcr commented on src/v/cluster/partition.cc:
```diff
@@ -571,13 +576,9 @@ partition::cloud_storage_timequery(storage::timequery_config cfg) {
-      _raft->ntp(),
-      cfg.time,
-      cfg.max_offset);
+    vlog(clusterlog.debug, "timequery (raft) ", _raft->ntp(), cfg);
```
weird that it didn't trigger an exception

andrwng commented on src/v/storage/types.cc:
```diff
@@ -93,8 +93,8 @@ std::ostream& operator<<(std::ostream& o, const timequery_result& a) {
 std::ostream& operator<<(std::ostream& o, const timequery_config& a) {
-    o << "{max_offset:" << a.max_offset << ", time:" << a.time
-      << ", type_filter:";
+    o << "{min_offset: " << a.min_offset << ", max_offset:" << a.max_offset
```
nit: maybe remove a space to keep spacing consistent?

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -2335,7 +2335,8 @@ disk_log_impl::make_reader(timequery_config config) {
     return _lock_mngr.range_lock(config).then(
       [this, cfg = config](std::unique_ptr<lock_manager::lease> lease) {
-          auto start_offset = _start_offset;
+          auto start_offset = cfg.min_offset;
```
Wondering if it makes sense to `std::max(_start_offset,...)` here too, given the scheduling point from taking the lock.

andrwng replied:
I don't think it matters since the range should only relevant segments anyway, regardless of start offset.

nvartolomei replied:
I don't think we should, and this makes the code better in my opinion. The `cfg.min_offset` here acts as a  optimistic concurrency control mechanism. If the start offset did advance then our assumptions are not valid anymore. So we throw in the reader. The exception bubbles up to the client which, will retry the request.

If we clamp the start offset then the caller must be very careful to check whether the log advanced because data was moved to cloud or because there was a log truncation. And what if it forgets about doing that?

A bit opinionated and half-baked but I think down the road this will lead to better code.

andrwng replied:
Yeah I'm fine with this. As a reader using the config as is seems less surprising

andrwng commented on src/v/kafka/server/replicated_partition.cc:
```diff
@@ -331,6 +331,9 @@ replicated_partition::timequery(storage::timequery_config cfg) {
         co_return res;
     }
+
+    // Local storage already handles this correctly. After we implement
+    // min_offset clamping for tiered storage, we can remove this code.
```
nit: thanks for this comment, but it's not quite clear what handling this is referring to, and what code this referring to. Mind clarifying?

nvartolomei replied:
I'm dropping this commit. Will update it in a different PR that is currently cooking.

andrwng commented on src/v/storage/log_reader.cc:
```diff
@@ -533,23 +533,27 @@ bool log_reader::is_done() {
-              if (record_t >= t) {
-                  auto record_o = model::offset{r.offset_delta()}
-                                  + b.base_offset();
+              if (
+                record_t >= t && min_offset <= record_o
+                && record_o <= max_offset) {
```
👍 

I wonder if we could end up never returning stop::yes if the timestamp we're looking for is in between batches.

nvartolomei replied:
@andijcr I can't imagine a case where that condition would be hit in practical terms so I don't want to add it. The code is correct as-is so I'm avoiding that "premature optimization" :)

andijcr replied:
was thinking more along the lines of separating the range of the search vs the predicate of the search, but you added `struct offset_range` which is overall better

andrwng replied:
After rereading this code, my suggestion doesn't make sense. Sorry for the noise

andrwng reviewed (commented):
Great find!

Probably worth updating the PR title to reflect that this is a bug fix

Also should this be backported to 23.2 as well? IIRC DeleteRecords support landed then

nvartolomei commented on src/v/storage/lock_manager.cc:
```diff
@@ -44,8 +44,9 @@ lock_manager::range_lock(const timequery_config& cfg) {
-          return s->offsets().get_base_offset() <= cfg.max_offset;
+          // Must be within offset bounds too.
+          return cfg.min_offset <= s->offsets().get_dirty_offset()
+                 && s->offsets().get_base_offset() <= cfg.max_offset;
```
Tried to explain it. Let me know if this works or no. More than happy to add extra details!

nvartolomei commented:
* Rebased on dev to resolve conflicts after a PR introduced by Willem to fix an unrelated timequery bug.
* Addressed reviewer's comments.
* Updated offset_range to bounded_offset_range to avoid misuse. It is less useful and more useful both at the same time!

Let's see what CI says.

andrwng commented on src/v/storage/log_reader.h:
```diff
@@ -259,15 +259,34 @@ class log_reader final : public model::record_batch_reader::impl {
 
 /**
- * Assuming caller has already determined that this batch contains
+ * Assuming caller has already determined that this batch contains (!)
```
nit: what is the (!) is referring to?

nvartolomei replied:
I was trying to draw attention to this line :) Added the word "critical" to it too.

andrwng commented on src/v/cluster/partition.cc:
```diff
@@ -513,7 +513,12 @@ partition::timequery(storage::timequery_config cfg) {
         // precise, so once we query we might still fall back to cloud
         // storage
-        auto result = co_await local_timequery(cfg);
+        // We also need to adjust the lower bound for the local query.
+        auto local_query_cfg = cfg;
+        local_query_cfg.min_offset = std::max(
+          log()->from_log_offset(_raft->start_offset()),
+          local_query_cfg.min_offset);
+        auto result = co_await local_timequery(local_query_cfg);
```
Could you write a commit message for this commit? Without more context, it's easy to thinking that there are no functional changes in this commit.

nvartolomei replied:
https://github.com/redpanda-data/redpanda/pull/18112/commits/f0514344c3b0672783041b782830c4b92c43681e

nvartolomei commented:
[Force pushed](https://github.com/redpanda-data/redpanda/compare/fe96ed21f79392b3b69482ba6546c2d568279d82..7214d2f26f869ce3b45e5428df370715745a3db7):
* improve [commit message](https://github.com/redpanda-data/redpanda/pull/18112/commits/f0514344c3b0672783041b782830c4b92c43681e) as requested by @andrwng
* add a fix for an edge case where cloud storage shouldn't be read at all  https://github.com/redpanda-data/redpanda/pull/18112/commits/f906e2480a47194242286a606b46389479f896fa
* commented out trim prefix with tiered storage as they run into an (existing) edge case which will be addressed in another PR

nvartolomei commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/2369538ecf81ffce8321d4b83a2f229a3cda72a3..ed192027a2e6f37170ae57a6702e2b5a118e1612):
* Fix off-by-one error in reader max offset
* Rename bounded_offset_range to bounded_offset_interval and redesign the API to make it easier to use correctly/harder to misuse

nvartolomei commented:
Last 2 force-pushes fixes some typos in text.

andrwng commented on tests/rptest/tests/timequery_test.py:
```diff
@@ -458,6 +458,66 @@ def query_slices(tid):
         assert not any([e > 0 for e in errors])
 
+    @cluster(num_nodes=4)
+    # @parametrize(cloud_storage=True, spillover=False)
+    # @parametrize(cloud_storage=True, spillover=True)
```
nit: remove?

nvartolomei replied:
I will re-enable this in the next PR https://github.com/redpanda-data/redpanda/pull/18097

andrwng commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -132,7 +132,7 @@ static ss::future<list_offset_partition_response> list_offsets_partition(
       kafka_partition->start_offset(),
       timestamp,
-      offset,
+      model::prev_offset(offset),
```
Yikes, nice catch!

andrwng replied:
I'm wondering what the implications of this are. Presumably if the `offset` doesn't exist, the timequery will just return early, and if it does exist, it won't affect the result of a timequery. Is that right?

Maybe a good idea to add any implications (or lackthereof) into the commit message

nvartolomei replied:
> and if it does exist, it won't affect the result of a timequery. Is that right?

You mean prior to this commit? It would have returned a potentially non-committed offset even with acks=all. Normally, redpanda doesn't expose such offsets unless write caching is in use. Fetch would have failed with out of range too, causing the consumer to reset.

andrwng reviewed (approved):
Some unit testing would be nice for some of the lower level functions, but otherwise this pretty much LGTM

nvartolomei commented:
Merging this to unblock https://github.com/redpanda-data/redpanda/pull/18097. Will address comments as follow ups.

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
/backport v23.3.x

vbotbuildovich commented:
Failed to create a backport PR to v23.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-18112-v23.3.x-984 remotes/upstream/v23.3.x
git cherry-pick -x 99d2bec5f7ee765cb3de88b446278262f6dae84f d97d61fb8a9b06fbde9dcf7ff03799bf200b561b 4f87afa392201c493f24f21af3e1cd7f0727649f f13bfa6c490490487d9a926c9a5d4e441adc3ca6 76a1ea2452b09a5730f2574646fc06ab2b8b8e32 f9ed5cabe479b355d370bec6bc9b693ad2928f3c a40999d2a09e0586c3fa81521c4d5fb5d0abc9dc 8f2de964c0e915f4f10ae8eb74400e6288c5680f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/8984732774).

nvartolomei commented on src/v/model/offset_interval.h:
```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Redpanda Data, Inc.
```
When reader config is reused the bounds can be reset too. It would be done only once, when the reader is created. By once I mean that is done once per read "make reader" operation.

nvartolomei commented on src/v/cluster/partition.cc:
```diff
@@ -506,7 +506,9 @@ partition::timequery(storage::timequery_config cfg) {
       = may_read_from_cloud()
-        && _cloud_storage_partition->bounds_timestamp(cfg.time);
+        && _cloud_storage_partition->bounds_timestamp(cfg.time)
+        && cfg.min_offset < kafka::offset_cast(
+             _cloud_storage_partition->next_kafka_offset());
```
I will think about it. I'm proposing to remove the masking code in the [next PR](https://github.com/redpanda-data/redpanda/pull/18097).

dotnwat commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -132,7 +132,7 @@ static ss::future<list_offset_partition_response> list_offsets_partition(
       kafka_partition->start_offset(),
       timestamp,
-      offset,
+      model::prev_offset(offset),
```
> kafka: fix off-by-one in timequery
> Not easy to test that this is right so not going to for now.

@nvartolomei please include in commit message more detail. the commit by inspection implies an off-by-one error, but there are critically useful pieces of information a reader in the future may want. for example, how was it discovered, what are symptoms of it being wrong, etc...

nvartolomei commented on src/v/model/offset_interval.h:
```diff
@@ -0,0 +1,80 @@
+        return {min, max};
+    }
+
+    inline bool overlaps(const bounded_offset_interval& other) const noexcept {
+        return _min <= other._max && _max >= other._min;
+    }
+
+    inline bool contains(model::offset o) const noexcept {
```
TIL

nvartolomei commented on src/v/model/offset_interval.h:
```diff
@@ -0,0 +1,80 @@
+    bounded_offset_interval(model::offset min, model::offset max) noexcept
+      : _min(min)
+      , _max(max) {
+#ifndef NDEBUG
```
My intention is to use this in hot-loops and I made the assumption that it will be impactful enough.

E.g. I want checks like the one in 

https://github.com/redpanda-data/redpanda/pull/18112/files#diff-adc3b4388ce88f3bd5b5de62a42ebc298737e9da94fa94ea9cde18342e246435R48 

or

https://github.com/redpanda-data/redpanda/blob/76a1ea2452b09a5730f2574646fc06ab2b8b8e32/src/v/storage/log_reader.cc#L558

to be compiled to nothing more than a single if statement.
